org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/CompilationUnitDeclaration.java

```diff
@@ -75,6 +75,7 @@ public int compare(Object o1, Object o2) {
 	public TypeDeclaration[] types;
 	public ModuleDeclaration moduleDeclaration;
 	public int[][] comments;
+	public UnnamedClass unnamedClass;
```

Comment:
Did you describe somewhere why the unnamedClass is not stored in the types array as the single element?

Reply:
It's not obvious as per spec and grammar and javac results that the unnamedClass is always the single type element in the compilation AST. There are counter examples such as

file A.java
```java
void main() {}
class A{}
```
which would even showcase that javac here sees 2 root types in the compilation unit: the unnamed one (resulting in A.class) and the named `A`, and this case results in an error because `A` is already defined.

Also, we may need to directly check the existence of an `unnamedClass` from time to time as it affects further compilation steps, so keeping a particular ref has been convenient.
But I think I see your point and we can instead of a field make it a method
```java
UnnamedClass unnamedClass() {
  return types == null ? null : Stream.of(types).filter(UnnamedClass.class::isInstance).map(UnnamedClass.class::cast).findAny().orElse(null);
}
```
Is this what you would recommend here?

Reply:
> Is this what you would recommend here?

Firstly I wanted to understand the reasoning behind it. 

Generally speaking, it would be nice if the approach wasn't diverging too much from CompilationUnitDeclaration.isPackageInfo() or CompilationUnitDeclaration.isModuleInfo()

https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-classes-instance-main-methods-jls.html to me reads as if it's always the single top-level class in the CU. Which parts do you find ambiguous? Maybe that should be discussed before making a decision about the AST?

Reply:
The ambiguity is in the grammar, when parsing. The example I gave earlier
```java
// A.java
void main() {}
class A{}
```
Is ambiguous at parsing: is A a top-level A or a nested A$A ? This is something we cannot easily sort when parsing and we need further steps later to decide that (according to the spec and javac, it's clearly A$A here, but it's not trivial to parse it).

CompilationUnit.isPackageInfo() or CompilationUnit.isModuleInfo() are relatively simple compared because they do use different parsing rules very early, so they leave less room to parsing ambiguity.

> Maybe that should be discussed before making a decision about the AST?

Guess what I'm trying with this PR ;) I'm really willing to discuss it and get a sufficient AST in place ASAP so we can then build on top of it.
If I read correctly, you're expecting `CompilationUnitDeclaration.isUnnamedClass()`. If so, what do you think such method would bring over my current proposal (assuming I remove the explicit unnamedClass field to replace with a method resolving against the resolved types) ?

Reply:
I’d expect the compilation unit to contain exactly one top-Level type declaration if it’s an unnamed class. What you suggested with the stream based boolean method indicted that there may be more than one top-level. I’d certainly expect that this is never the case so `types!=null && types.length ==1 && types[0] instanceof UnnamedClass` would be the condition that should return the correct result. Is that how you’d model it? (Its not clear to me yet from this WIP PR)

Reply:
> What you suggested with the stream based boolean method indicted that there may be more than one top-level

Indeed, the current AST model may allow extra top-level types together with unnamed class, but previous steps (Parser.consumeInternalCompilationUnitWithPotentialUnnamedClass) and next steps (Scope) makes that such AST cannot be fully compiled with usual ways. I can add some checks in CompilationUnitDeclaration.analyzeCode to check things before trying to resolve. Would that be enough?

Reply:
What are the implications of what you just said? As far as I understand it, all code that processes the types would need to be aware of the wrong nesting of the type declarations and all code that works with the unnamed class declaration would need to be aware of the fact that some of its siblings are in fact contained? This doesn’t sound great as a model to start with. 

Would be nice to get the opinion of people that are more familiar with the compilers code base than I am.

Reply:
> As far as I understand it, all code that processes the types would need to be aware of the wrong nesting of the type declarations and all code that works with the unnamed class declaration would need to be aware of the fact that some of its siblings are in fact contained? 

There is no wrong nesting if the AST was built by the Parser. It's just that someone using the AST can put another top-level type in the `types` array, and will get reported problems later if they do so.
Feel free to suggest concrete additional test-cases we can add to cover your concerns, it will help in finding the right problem and solutions.

Reply:
> Indeed, the current AST model may allow extra top-level types together with unnamed class,

maybe I misunderstood. You’re saying the parser will never create a wrongly nested AST? Or would it?

Reply:
when an unnamed class is detected (ie a method or a field is present at the root of the unit), then all the other types defined at the root of the compilation unit then become member types nested in the unnamed class.